## 1. Evaluation initiale sur le dataset Adult

### 1.1 Chargement et exécution des modèles sur le dataset Adult
Elle charge les données Adult et exécute les trois modèles (XGBoost, BRCG, RIPPER) pour comparer leurs performances.

In [3]:
import import_ipynb  
import brcg_model
import adultdata_preparation_evaluat
import xgboost_rule_extraction
import explainer
import pandas as pd


# Charger les données
x_train, x_test, y_train, y_test, mapping = adultdata_preparation_evaluat.prepare_adult_dataset()
print("Données chargées depuis le notebook principal") 

# Nettoyer les données                          
x_train = x_train.drop(columns=['label'], errors='ignore')
x_test = x_test.drop(columns=['label'], errors='ignore')

# Exécuter le notebook XGBoost - cela va définir train_xgboost dans l'espace de noms
%run ./xgboost_rule_extraction.ipynb

# Vérifier que train_xgboost est maintenant disponible
print("train_xgboost disponible:", 'train_xgboost' in dir())

if 'train_xgboost' in dir():
    # Utiliser XGBoost
    
    x_test = x_test.drop(columns=['label'], errors='ignore')
    xgb_model, y_pred_xgb, xgb_metrics = xgboost_rule_extraction.train_xgboost(
        x_train, y_train, x_test, y_test
    )
    print("Métriques XGBoost:", xgb_metrics)
else:
    print("ERREUR: train_xgboost n'a pas été chargé correctement")
    print("Contenu de l'espace de noms:", [v for v in dir() if 'train' in v])

# BRCG
brcg_model_, fb_brcg, columns_brcg, train_time_brcg, brcg_metrics, brcg_rules = brcg_model.train_brcg(
    x_train, y_train, x_test, y_test
)
print("Métriques BRCG:", brcg_metrics)

# RIPPER
from ripper_model import train_ripper
ripper_model_, y_pred_ripper, ripper_metrics, ripper_rules = train_ripper(
    x_train, y_train, x_test, y_test
)
print("Métriques RIPPER:", ripper_metrics)

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.2181
Iteration: 2, Objective: 0.1897
Iteration: 3, Objective: 0.1854
Iteration: 4, Objective: 0.1843
Iteration: 5, Objective: 0.1843
Iteration: 6, Objective: 0.1843
Résultats BRCG
Accuracy: 0.8189774297558728
F1-score: 0.8026137894465102
Precision: 0.8063169426792735
Recall: 0.8189774297558728
Impossible d'afficher les règles


100%|██████████| 14/14 [00:00<00:00, 16.77it/s]


Total Support: 23.48% (6116)
After removing duplicates:
Unique Coverage: 21.88% (5698)


100%|██████████| 723/723 [00:00<00:00, 3600.70it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using Majority class to predict non-covered cases.
F1-score (if using entire dataset): 0.7775


100%|██████████| 723/723 [00:00<00:00, 3989.28it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using model to predict non-covered cases.
F1-score (if using entire dataset): 0.8689
Données chargées depuis le notebook principal


100%|██████████| 14/14 [00:00<00:00, 18.89it/s]


Total Support: 23.48% (6116)
After removing duplicates:
Unique Coverage: 21.88% (5698)


100%|██████████| 723/723 [00:00<00:00, 3951.97it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using Majority class to predict non-covered cases.
F1-score (if using entire dataset): 0.7775


100%|██████████| 723/723 [00:00<00:00, 3643.01it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using model to predict non-covered cases.
F1-score (if using entire dataset): 0.8689
train_xgboost disponible: True
Métriques XGBoost: {'accuracy': 0.8753262705358513, 'balanced_accuracy': 0.7969251078262138, 'precision': 0.7990543735224587, 'recall': 0.645448758752387, 'f1': 0.7140845070422536, 'accuracy_train': 0.8748080466830467, 'f1_train': 0.7087092451987494, 'training_time': 0.4665820598602295}
Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.2181
Iteration: 2, Objective: 0.1897
Iteration: 3, Objective: 0.1854
Iteration: 4, Objective: 0.1843
Iteration: 5, Objective: 0.1843
Iteration: 6, Objective: 0.1843
Métriques BRCG: {'accuracy': 0.8189774297558728, 'f1': 0.8026137894465102, 'pr

### 1.2 Comparaison des performances des modèles

In [4]:
print("\n===== Comparaison des modèles =====")

print(f"XGBoost  -> Accuracy: {xgb_metrics['accuracy']:.4f} | F1-score: {xgb_metrics['f1']:.4f}")
print(f"BRCG     -> Accuracy: {brcg_metrics['accuracy']:.4f} | F1-score: {brcg_metrics['f1']:.4f}")
print(f"RIPPER   -> Accuracy: {ripper_metrics['accuracy']:.4f} | F1-score: {ripper_metrics['f1']:.4f}")


===== Comparaison des modèles =====
XGBoost  -> Accuracy: 0.8753 | F1-score: 0.7141
BRCG     -> Accuracy: 0.8190 | F1-score: 0.8026
RIPPER   -> Accuracy: 0.8188 | F1-score: 0.8009


--> XGBoost surpasse en accuracy mais souffre d'un déséquilibre de classes (F1-score plus faible)

--> RIPPER offre le meilleur compromis avec un F1-score élevé et une bonne accuracy

--> BRCG est très proche de RIPPER en performance

## 2. Évaluation multicritères sur les 5 datasets (BRCG RIPPER)

### 2.1 Import des bibliothèques pour l'évaluation 

In [5]:
import pandas as pd
import numpy as np
import time
import psutil
import os
import gc
from tqdm import tqdm


import import_ipynb
import ripper_model
import brcg_model
# Import des modèles séparés



from aix360.algorithms.rbm import FeatureBinarizer, BooleanRuleCG
import wittgenstein as lw

import os
os.makedirs("results", exist_ok=True)

In [ ]:
import warnings
warnings.filterwarnings("ignore")

import matplotlib.pyplot as plt
import traceback

from brcg_model import train_brcg, evaluate_brcg
from ripper_model import train_ripper, evaluate_ripper


from cmr_model import run_cmr

n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=20000, dataset_name="", iteration=0):
    results = {}

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]

    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

    # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = train_brcg(
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        ram = psutil.Process().memory_info().rss / 1024**2

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            train_time_brcg,
            ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"\nErreur BRCG sur le dataset {dataset_name}, itération {iteration}: {e}")
        traceback.print_exc()
        results["BRCG"] = None

    # -------- RIPPER --------
    try:
        model_ripper, y_pred_ripper, metrics, rules = train_ripper(
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        ram = psutil.Process().memory_info().rss / 1024**2

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            metrics["train_time"],
            ram
        )

    except Exception as e:
        print(f"\nErreur RIPPER sur le dataset {dataset_name}, itération {iteration}: {e}")
        traceback.print_exc()
        results["RIPPER"] = None

    
    # -------- CMR --------
    try:
        results["CMR"] = run_cmr(
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

    except Exception as e:
        print(f"\nErreur CMR sur le dataset {dataset_name}, itération {iteration}: {e}")
        traceback.print_exc()
        results["CMR"] = None

    return results

datasets = {
    "KDD99": "../Datasets/KDD99/",
    "BIG15": "../Datasets/BIG15/",
    "DoH20": "../Datasets/DoH20/",
    "UNSW-NB15": "../Datasets/UNSW-NB15/processed/",
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {ds: [] for ds in datasets.keys()}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        if name == "CIC2017":
            X_train = pd.read_parquet(path + "X_train.parquet")
            X_test = pd.read_parquet(path + "X_test.parquet")
            y_train = pd.read_csv(path + "y_train.csv").squeeze()
            y_test = pd.read_csv(path + "y_test.csv").squeeze()

        elif name == "UNSW-NB15":
            train = pd.read_csv(path + "le_train.csv")
            test = pd.read_csv(path + "le_test.csv")

            X_train = train.drop(["id", "attack_cat", "label"], axis=1)
            y_train = train["label"]

            X_test = test.drop(["id", "attack_cat", "label"], axis=1)
            y_test = test["label"]

        else:
            X_train = pd.read_csv(path + "X_train.csv")
            X_test = pd.read_csv(path + "X_test.csv")
            y_train = pd.read_csv(path + "y_train.csv").squeeze()
            y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=20000,
                dataset_name=name,
                iteration=it+1
            )
            results_all_iterations[name].append(res)

          
            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = [name, method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations.csv", index=False)

        del X_train, X_test, y_train, y_test
        if "train" in locals():
            del train
        if "test" in locals():
            del test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds in datasets.keys():
    runs = results_all_iterations.get(ds, [])

 
    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = [ds, method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif.csv", index=False)

if os.path.exists("results/progress_iterations.csv"):
    df_progress = pd.read_csv("results/progress_iterations.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}.png")
        plt.close()

print(df_results)

  0%|          | 0/5 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0370
Iteration: 2, Objective: 0.0283
Iteration: 3, Objective: 0.0246
Iteration: 4, Objective: 0.0236


 20%|██        | 1/5 [01:54<07:38, 114.66s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0268
Iteration: 2, Objective: 0.0268


 40%|████      | 2/5 [05:52<09:21, 187.14s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 60%|██████    | 3/5 [10:50<07:55, 237.85s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1920


 80%|████████  | 4/5 [16:28<04:37, 277.22s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0747
Iteration: 2, Objective: 0.0605
Iteration: 3, Objective: 0.0605


100%|██████████| 5/5 [41:01<00:00, 492.24s/it]


      Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
3       BIG15    BRCG  0.984057  0.967507             NaN           0.98405   
4       BIG15  RIPPER  0.990860  0.986832             NaN           0.99085   
5       BIG15     CMR  0.997951  0.992506           95.48           1.00000   
12    CIC2017    BRCG  0.944590  0.941422             NaN           0.94800   
13    CIC2017  RIPPER  0.997398  0.995925             NaN           0.99740   
14    CIC2017     CMR  0.998950  0.997434           92.88           1.00000   
6       DoH20    BRCG  0.961763  0.382162             NaN           0.96585   
7       DoH20  RIPPER  0.983330  0.978983             NaN           0.98255   
8       DoH20     CMR  1.000000  0.999665          100.00           1.00000   
0       KDD99    BRCG  0.987422  0.984609             NaN           0.98745   
1       KDD99  RIPPER  0.998400  0.997252             NaN           0.99840   
2       KDD99     CMR  0.999750  0.998145           

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

In [ ]:

import warnings
warnings.filterwarnings("ignore")

import matplotlib.pyplot as plt
import traceback

from brcg_model import train_brcg, evaluate_brcg
from ripper_model import train_ripper, evaluate_ripper


from cmr_model import run_cmr

n_iterations = 10
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=20000, dataset_name="", iteration=0):
    results = {}

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]

    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

    # -------- BRCG --------
   
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = train_brcg(
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        ram = psutil.Process().memory_info().rss / 1024**2

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            train_time_brcg,
            ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"\nErreur BRCG sur le dataset {dataset_name}, itération {iteration}: {e}")
        traceback.print_exc()
        results["BRCG"] = None

    # -------- RIPPER --------
    try:
        model_ripper, y_pred_ripper, metrics, rules = train_ripper(
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        ram = psutil.Process().memory_info().rss / 1024**2

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            metrics["train_time"],
            ram
        )

    except Exception as e:
        print(f"\nErreur RIPPER sur le dataset {dataset_name}, itération {iteration}: {e}")
        traceback.print_exc()
        results["RIPPER"] = None


    # -------- CMR --------
    try:
        results["CMR"] = run_cmr(
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

    except Exception as e:
        print(f"\nErreur CMR sur le dataset {dataset_name}, itération {iteration}: {e}")
        traceback.print_exc()
        results["CMR"] = None

    return results

datasets = {
    "KDD99": "../Datasets/KDD99/",
    "BIG15": "../Datasets/BIG15/",
    "DoH20": "../Datasets/DoH20/",
    "UNSW-NB15": "../Datasets/UNSW-NB15/processed/",
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {ds: [] for ds in datasets.keys()}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        if name == "CIC2017":
            X_train = pd.read_parquet(path + "X_train.parquet")
            X_test = pd.read_parquet(path + "X_test.parquet")
            y_train = pd.read_csv(path + "y_train.csv").squeeze()
            y_test = pd.read_csv(path + "y_test.csv").squeeze()

        elif name == "UNSW-NB15":
            train = pd.read_csv(path + "le_train.csv")
            test = pd.read_csv(path + "le_test.csv")

            X_train = train.drop(["id", "attack_cat", "label"], axis=1)
            y_train = train["label"]

            X_test = test.drop(["id", "attack_cat", "label"], axis=1)
            y_test = test["label"]

        else:
            X_train = pd.read_csv(path + "X_train.csv")
            X_test = pd.read_csv(path + "X_test.csv")
            y_train = pd.read_csv(path + "y_train.csv").squeeze()
            y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=20000,
                dataset_name=name,
                iteration=it+1
            )
            results_all_iterations[name].append(res)

           
            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = [name, method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations.csv", index=False)

        del X_train, X_test, y_train, y_test
        if "train" in locals():
            del train
        if "test" in locals():
            del test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds in datasets.keys():
    runs = results_all_iterations.get(ds, [])

    
    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = [ds, method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif.csv", index=False)

if os.path.exists("results/progress_iterations.csv"):
    df_progress = pd.read_csv("results/progress_iterations.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}.png")
        plt.close()

print(df_results)

  0%|          | 0/5 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0370
Iteration: 2, Objective: 0.0283
Iteration: 3, Objective: 0.0246
Iteration: 4, Objective: 0.0236


100%|██████████| 38/38 [00:00<00:00, 57.97it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0375
Iteration: 2, Objective: 0.0248
Iteration: 3, Objective: 0.0221
Iteration: 4, Objective: 0.0211


100%|██████████| 38/38 [00:00<00:00, 69.18it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0383
Iteration: 2, Objective: 0.0237
Iteration: 3, Objective: 0.0212
Iteration: 4, Objective: 0.0207


100%|██████████| 39/39 [00:00<00:00, 69.11it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0359
Iteration: 2, Objective: 0.0276
Iteration: 3, Objective: 0.0244
Iteration: 4, Objective: 0.0236


100%|██████████| 37/37 [00:00<00:00, 55.54it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0395
Iteration: 2, Objective: 0.0254
Iteration: 3, Objective: 0.0218
Iteration: 4, Objective: 0.0218


100%|██████████| 39/39 [00:00<00:00, 68.22it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0382
Iteration: 2, Objective: 0.0281
Iteration: 3, Objective: 0.0253


100%|██████████| 37/37 [00:00<00:00, 60.04it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0383
Iteration: 2, Objective: 0.0251
Iteration: 3, Objective: 0.0223
Iteration: 4, Objective: 0.0212
Iteration: 5, Objective: 0.0212


100%|██████████| 39/39 [00:00<00:00, 68.86it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0365
Iteration: 2, Objective: 0.0224
Iteration: 3, Objective: 0.0208
Iteration: 4, Objective: 0.0203


100%|██████████| 38/38 [00:00<00:00, 68.11it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0384
Iteration: 2, Objective: 0.0248
Iteration: 3, Objective: 0.0218
Iteration: 4, Objective: 0.0204
Iteration: 5, Objective: 0.0204


100%|██████████| 39/39 [00:00<00:00, 71.23it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0357
Iteration: 2, Objective: 0.0239
Iteration: 3, Objective: 0.0205
Iteration: 4, Objective: 0.0196


 20%|██        | 1/5 [18:20<1:13:20, 1100.22s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0274
Iteration: 2, Objective: 0.0271


100%|██████████| 54/54 [00:02<00:00, 18.36it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0264
Iteration: 2, Objective: 0.0242


100%|██████████| 54/54 [00:02<00:00, 18.93it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0285
Iteration: 2, Objective: 0.0285


100%|██████████| 54/54 [00:03<00:00, 16.14it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0254
Iteration: 2, Objective: 0.0254


100%|██████████| 54/54 [00:02<00:00, 18.91it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0278
Iteration: 2, Objective: 0.0276


100%|██████████| 54/54 [00:02<00:00, 18.87it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0275
Iteration: 2, Objective: 0.0274


100%|██████████| 54/54 [00:03<00:00, 16.57it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0271
Iteration: 2, Objective: 0.0271


100%|██████████| 54/54 [00:02<00:00, 18.23it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0264
Iteration: 2, Objective: 0.0264


100%|██████████| 54/54 [00:02<00:00, 18.59it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0260
Iteration: 2, Objective: 0.0260


100%|██████████| 54/54 [00:02<00:00, 18.70it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0273
Iteration: 2, Objective: 0.0273


 40%|████      | 2/5 [51:55<1:21:55, 1638.65s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:09<00:00,  2.95it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:10<00:00,  2.81it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:09<00:00,  2.95it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:09<00:00,  2.97it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:09<00:00,  2.93it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:09<00:00,  2.95it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:09<00:00,  2.95it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:10<00:00,  2.84it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [00:10<00:00,  2.87it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 60%|██████    | 3/5 [1:31:45<1:06:02, 1981.40s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1944
Iteration: 2, Objective: 0.1302


100%|██████████| 42/42 [00:03<00:00, 11.85it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1921
Iteration: 2, Objective: 0.1275


100%|██████████| 42/42 [00:03<00:00, 11.59it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1902
Iteration: 2, Objective: 0.1289


100%|██████████| 42/42 [00:03<00:00, 11.89it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1882
Iteration: 2, Objective: 0.1316


100%|██████████| 42/42 [00:03<00:00, 11.41it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1885
Iteration: 2, Objective: 0.1244


100%|██████████| 42/42 [00:03<00:00, 11.85it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1880
Iteration: 2, Objective: 0.1249


100%|██████████| 42/42 [00:03<00:00, 11.61it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1918
Iteration: 2, Objective: 0.1284


100%|██████████| 42/42 [00:03<00:00, 11.20it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1950
Iteration: 2, Objective: 0.1312


100%|██████████| 42/42 [00:07<00:00,  5.48it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1861


100%|██████████| 42/42 [00:03<00:00, 11.52it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.1865
Iteration: 2, Objective: 0.1269


 80%|████████  | 4/5 [2:19:11<38:43, 2323.05s/it]  

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0832
Iteration: 2, Objective: 0.0665
Iteration: 3, Objective: 0.0665


100%|██████████| 70/70 [00:06<00:00, 10.15it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0810
Iteration: 2, Objective: 0.0636
Iteration: 3, Objective: 0.0621


100%|██████████| 68/68 [00:07<00:00,  9.65it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0788
Iteration: 2, Objective: 0.0640
Iteration: 3, Objective: 0.0618


100%|██████████| 66/66 [00:07<00:00,  9.07it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0768
Iteration: 2, Objective: 0.0600
Iteration: 3, Objective: 0.0600
Iteration: 4, Objective: 0.0583


100%|██████████| 68/68 [00:07<00:00,  9.39it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0770
Iteration: 2, Objective: 0.0596
Iteration: 3, Objective: 0.0593


100%|██████████| 68/68 [00:07<00:00,  8.61it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0763
Iteration: 2, Objective: 0.0600
Iteration: 3, Objective: 0.0592
Iteration: 4, Objective: 0.0588


100%|██████████| 68/68 [00:07<00:00,  8.56it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0784
Iteration: 2, Objective: 0.0638
Iteration: 3, Objective: 0.0613


100%|██████████| 70/70 [00:07<00:00,  9.63it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0770
Iteration: 2, Objective: 0.0611
Iteration: 3, Objective: 0.0611


100%|██████████| 68/68 [00:07<00:00,  8.54it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0777
Iteration: 2, Objective: 0.0622
Iteration: 3, Objective: 0.0607


100%|██████████| 68/68 [00:08<00:00,  7.91it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0721
Iteration: 2, Objective: 0.0576
Iteration: 3, Objective: 0.0569
Iteration: 4, Objective: 0.0568


100%|██████████| 5/5 [5:53:55<00:00, 4247.06s/it]


      Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
3       BIG15    BRCG  0.983050  0.962613             NaN          0.983030   
4       BIG15  RIPPER  0.991645  0.985958             NaN          0.991640   
5       BIG15     CMR  0.997740  0.992278          95.300          1.000000   
12    CIC2017    BRCG  0.944123  0.940985             NaN          0.947805   
13    CIC2017  RIPPER  0.997207  0.995879             NaN          0.997210   
14    CIC2017     CMR  0.998760  0.997508          92.695          1.000000   
6       DoH20    BRCG  0.959914  0.414521             NaN          0.964390   
7       DoH20  RIPPER  0.987497  0.983250             NaN          0.987185   
8       DoH20     CMR  1.000000  0.999713          99.993          1.000000   
0       KDD99    BRCG  0.990002  0.986031             NaN          0.990020   
1       KDD99  RIPPER  0.998235  0.996902             NaN          0.998235   
2       KDD99     CMR  0.999750  0.998537          9

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

In [10]:
import sys
import sklearn
import numpy
import pandas

print("Python:", sys.version)
print("sklearn:", sklearn.__version__)
print("numpy:", numpy.__version__)
print("pandas:", pandas.__version__)

Python: 3.9.25 (main, Nov  3 2025, 22:44:01) [MSC v.1929 64 bit (AMD64)]
sklearn: 1.6.1
numpy: 1.23.5
pandas: 1.5.3


In [11]:
import platform
import psutil

print("OS :", platform.system(), platform.release())
print("CPU :", platform.processor())
print("RAM :", round(psutil.virtual_memory().total / (1024**3), 2), "GB")

OS : Windows 10
CPU : Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
RAM : 15.41 GB


Les expériences ont été réalisées dans l’environnement technique suivant :

Système d’exploitation : Windows 10
Processeur : Intel64 Family 6 Model 78 (GenuineIntel)
Mémoire RAM : 15.41 Go
Langage de programmation : Python 3.9.25 (64 bits)

Bibliothèques utilisées :

scikit-learn : 1.6.1
NumPy : 1.23.5
pandas : 1.5.3

Les expériences ont été exécutées en local sur cette machine.

Resume tableau 8 : Les résultats montrent qu'augmenter le paramètre de support minimal (s_min) réduit le nombre de règles extraites, ce qui diminue les taux de conflit mais aussi la couverture. 

Si les performances peuvent s'améliorer légèrement grâce à la réduction du bruit dans certains jeux de données, elles peuvent également se dégrader lorsque des motifs importants sont supprimés. Notamment, s_min = 1 et s_min = 3 produisent des résultats identiques, indiquant que toutes les règles extraites bénéficient déjà d'un support suffisant. Globalement, s_min = 3 offre un bon compromis en termes de couverture.…